File: bindings/julia/src/SafeMath.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# SPDX-FileCopyrightText: 2025 Hyperpolymath

"""
    SafeMath

Safe arithmetic operations that cannot crash or overflow unexpectedly.
"""
module SafeMath

export safe_div, safe_mod, safe_add, safe_sub, safe_mul, checked_add, checked_sub, checked_mul

"""
    safe_div(numerator::Integer, denominator::Integer) -> Union{Integer, Nothing}

Safely divide two integers, returning `nothing` on division by zero.
"""
function safe_div(numerator::T, denominator::T)::Union{T, Nothing} where T <: Integer
    denominator == 0 ? nothing : div(numerator, denominator)
end

"""
    safe_mod(numerator::Integer, denominator::Integer) -> Union{Integer, Nothing}

Safely compute modulo, returning `nothing` on division by zero.
"""
function safe_mod(numerator::T, denominator::T)::Union{T, Nothing} where T <: Integer
    denominator == 0 ? nothing : mod(numerator, denominator)
end

"""
    safe_add(a::Integer, b::Integer) -> Union{Integer, Nothing}

Safely add two integers, returning `nothing` on overflow.
"""
function safe_add(a::T, b::T)::Union{T, Nothing} where T <: Integer
    try
        Base.checked_add(a, b)
    catch e
        e isa OverflowError ? nothing : rethrow(e)
    end
end

"""
    safe_sub(a::Integer, b::Integer) -> Union{Integer, Nothing}

Safely subtract two integers, returning `nothing` on overflow.
"""
function safe_sub(a::T, b::T)::Union{T, Nothing} where T <: Integer
    try
        Base.checked_sub(a, b)
    catch e
        e isa OverflowError ? nothing : rethrow(e)
    end
end

"""
    safe_mul(a::Integer, b::Integer) -> Union{Integer, Nothing}

Safely multiply two integers, returning `nothing` on overflow.
"""
function safe_mul(a::T, b::T)::Union{T, Nothing} where T <: Integer
    try
        Base.checked_mul(a, b)
    catch e
        e isa OverflowError ? nothing : rethrow(e)
    end
end

"""
    checked_add(a::Integer, b::Integer) -> Integer

Add with overflow checking. Throws OverflowError on overflow.
"""
checked_add(a::T, b::T) where T <: Integer = Base.checked_add(a, b)

"""
    checked_sub(a::Integer, b::Integer) -> Integer

Subtract with overflow checking. Throws OverflowError on overflow.
"""
checked_sub(a::T, b::T) where T <: Integer = Base.checked_sub(a, b)

"""
    checked_mul(a::Integer, b::Integer) -> Integer

Multiply with overflow checking. Throws OverflowError on overflow.
"""
checked_mul(a::T, b::T) where T <: Integer = Base.checked_mul(a, b)

end # module

File: bindings/ocaml/lib/safe_math.ml
(* SPDX-License-Identifier: Apache-2.0 *)
(* SPDX-FileCopyrightText: 2025 Hyperpolymath *)

(** Safe mathematical operations with overflow detection. *)

let safe_div numerator denominator =
  if denominator = 0 then None
  else Some (numerator / denominator)

let safe_mod numerator denominator =
  if denominator = 0 then None
  else Some (numerator mod denominator)

let safe_add a b =
  (* Check for overflow using Int module bounds *)
  if b > 0 && a > max_int - b then None
  else if b < 0 && a < min_int - b then None
  else Some (a + b)

let safe_sub a b =
  (* Check for overflow *)
  if b < 0 && a > max_int + b then None
  else if b > 0 && a < min_int + b then None
  else Some (a - b)

let safe_mul a b =
  if a = 0 || b = 0 then Some 0
  else
    (* Check for overflow using division *)
    let would_overflow =
      if a > 0 then
        if b > 0 then a > max_int / b
        else b < min_int / a
      else
        if b > 0 then a < min_int / b
        else a <> 0 && b < max_int / a
    in
    if would_overflow then None
    else Some (a * b)
